### PR TITLE
WELCOME ALERT HISTORY SCREEN

### PR DESCRIPTION
**Pull request by rinslow**
Screenshot from 2020-04-06 22-25-45
Screenshot from 2020-04-06 22-26-02
Screenshot from 2020-04-06 22-26-10

# Summary 
Alert history screen is done! And it works perfectly.
The alert screen shows previous alerts, and marks new alerts in bright red colour. it supports scrolling down and up. The refresh button is disabled by default until a relevant* new `Alert` comes along, which then makes it turn on to bright purple.

When the user clicks the refresh button, the history snapshot updates and he is scrolled to the top automatically. 

This was done with intents of minimal intrusion to the Doctor's experience.

* An alert that is different from the last alert, or an alert that is the same but 15 minutes later or more.

# Solves
* https://github.com/Reznic/Inhalator/issues/18
* https://github.com/Reznic/Inhalator/issues/127

# Current coverage
59% 

# Performance
In order to fix previous performance issues that arose from a naive implementation approach, 
a more design-oriented approach was taken. 
The problem was that `AlertEntity` graphic objects were rendered and discarded multiple times a second, and while a normal Laptop can perform this task easily, the Raspberry Pi CPU failed to do so.  

# Design 
## Flyweight
In order to answer the performance issues that arose, I used the *Fly-weight* design pattern.
For that I had to separate the `AlertEntity` object to it's intrinsic state and extrinsic state, create a factory that would cache and return absent entities, and allow me to inject them with a state. 
This may sound a little complicated but the code is extremely simple, and well-documented.

## Refactors
* `Alert` class now has a `seen` property
* a new `AlertHistory` object is now a part of the `AlertQueue` 
* `alert.py` file has been splitted to multiple files. 
* `observer` renamed to `observable`
* No more `LockThresholdsButton`
* Importing `AlertCodes` directly everywhere

# Bug-fixes
* You can no longer enqueue an alert without a timestamp-context

# Configurations
* Added a new 'delta_time_between_alerts' to the `config.json` file, meaning how much time to wait between displaying the same alert on the history

# Functionalities Tested
* Alerts appear on the history screen
* Scrolling down works
* Scrolling up works
* Scrolling down doesn't work when it doesn't make sense to 
* Scrolling up doesn't work when it doesn't make sense to 
* Back button functionality
* Refresh button functionality 
* Seen and unseen alerts 
* Same alerts trigger a warning after some time 
* Same alerts don't trigger a warning pior to some time

Tested graphically on a Raspberry-Pi

### CODE REVIEW

File: data/alert.py [hunk 1]
```diff
@@ -49,14 +47,16 @@ class Alert(object):
         AlertCodes.NO_BATTERY: "No Battery",
     }
 
-    def __init__(self, alert_code, timestamp=None):
+    def __init__(self, alert_code, timestamp=None):  # TODO: Add 'Seen'
```
Comment (Reznic) [inline data/alert.py:50]: Is this Todo relevant?
You allready implemented 'seen'..

File: data/alert.py [hunk 2]
```diff
@@ -49,14 +47,16 @@ class Alert(object):
...
         self.code = alert_code
         if timestamp is None:
-            self.timestamp = time.time()
+            self.timestamp = time.time()  # TODO: Remove
```
Comment (Reznic) [inline data/alert.py:53]: Now you can remove that. right?
Now every new alert initialized with timestamp

Comment (rinslow) [reply data/alert.py:53]: Every `enqueue_alert()` demands a time. 
I think we'll do this in a future issue to not make this one too big

File: tests/integration/test_volume.py [hunk 3]
```diff
@@ -145,4 +145,4 @@ def test_sampler_alerts_when_volume_exceeds_maximum(events, measurements, config
     assert len(events.alerts_queue) > 0
 
     all_alerts = list(events.alerts_queue.queue.queue)
-    assert all(alert == alerts.AlertCodes.VOLUME_HIGH for alert in all_alerts)
\ No newline at end of file
+    assert all(alert == AlertCodes.VOLUME_HIGH for alert in all_alerts)
```
Comment (avrahamshukron) [inline tests/integration/test_volume.py:148]: Add newlie

File: data/alerts_queue.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
+from queue import Queue
+
+from data.alert import Alert, AlertCodes
+from data.history import AlertsHistory
+from data.observable import Observable
+
+
+class AlertsQueue(object):
+    MAXIMUM_ALERTS_AMOUNT = 2
+
```
[the remark this answers is not in the document]
Comment (rinslow) [reply data/alerts_queue.py:10]: yeah I definitely agree, it started out as a typo

File: data/history.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+    def __init__(self):
+        self.stack = deque(maxlen=40)
+        self.observable = Observable()
+        self.time_difference_between_same_alerts = 60 * 15  # 15 Minutes
+
+    def __len__(self):
+        return len(self.stack)
+
+    def copy(self):
```
[the remark this answers is not in the document]
Comment (rinslow) [reply data/history.py:18]: I'll document it

File: default_config.json [hunk 6]
```diff
@@ -35,5 +35,6 @@
   "log_enabled": true,
   "graph_seconds": 12.0,
   "low_battery_percentage": 15,
-  "mute_time_limit" : 120
+  "mute_time_limit" : 120,
+  "delta_time_between_alerts": 900
```
[the remark this answers is not in the document]
Comment (rinslow) [reply default_config.json:39]: Changed to 5 minutes following our private discussion

**Pull request merged**